### ISSUE TITLE
Mechanism for ephemeral updates

### ISSUE THREAD

**Issue by ept** [automerge/automerge#320]
Many collaborative apps need a way of sharing real-time updates with other users without persisting those updates. This happens when the real-time updates are useful for communication with other users, but have no long-term value. For example:

* Pixel-by-pixel updates while dragging an object (when the object is released, its final position should be persisted, but it is not necessary to store all the intermediate positions while the dragging is in progress)
* A user's current cursor position and selection in a text editor (if other users' cursors should be displayed)
* A user's current mouse pointer position in a graphical app (if other users' mouse pointers should be displayed)

In these situations, it would be possible to make all those changes part of a document, but it would be wasteful, since nobody is interested in the history of all those fine-grained updates, and there can easily be many updates per second.

One option is to leave such transient updates entirely out of Automerge and to communicate them via a separate real-time channel between users. However, this adds complexity to the app, since it needs to merge the persistent state with the real-time updates: in the object dragging example, the current position of an object may be determined either by that object's persistent state or the latest real-time update to that object's state, whichever is newer. When a dragging operation begins, the real-time updates start taking precedence over the persisted object position, but when the dragging is complete, the object's persistent position is updated and the real-time position updates are no longer valid. This potentially fiddly logic might better be implemented in a general-purpose library, so that the app doesn't have to deal with this complexity.

Another concern is that the currently proposed mechanism for cursors (#313) requires the cursor object to be part of an Automerge document in order to be able to convert between the cursor's elemId and its index. If cursors are only communicated out-of-band outside of Automerge, we cannot use this mechanism.

This suggests that there should be a way of making ephemeral updates to an Automerge document in a way that allows those updates to be shared over the network, but without persisting those updates. @pvh has suggested that one possible implementation would be to use separate actorIds for the persistent and the ephemeral changes, and allow ephemeral changes to depend on persistent changes, but not the other way round.

Another thing to keep in mind is that there may be interplay between ephemeral updates and the undo feature (#318). We probably don't want to retain undo history for ephemeral updates, but cursor positions sometimes need to be updated on undo: for example, in a text editor, standard behaviour is that when you undo an insertion/deletion, the focus is moved to the place where the undone edit occurred. Not yet sure how best to model that, but it's something to keep in mind.

**Comment by ept** [comment]
There is one simple (and reasonably efficient) way how ephemeral updates can be implemented today, without any API changes. Every client creates two Automerge instances for every document, rather than just one:

```js
let persistent = Automerge.init(), ephemeral = Automerge.init()
```

To make a persistent change, the client makes the change on the persistent document and then copies it over to the ephemeral one:

```js
persistent = Automerge.change(persistent, doc => doc.persist = 'this!')
const change = Automerge.getLastLocalChange(persistent)
ephemeral = Automerge.applyChanges(ephemeral, [change])
// send change over the network to all other clients
```

To make an ephemeral change, the client makes the change only on the ephemeral instance:

```js
ephemeral = Automerge.change(ephemeral, doc => doc.forget = 'me please')
const change = Automerge.getLastLocalChange(ephemeral)
// send change over the network to all other clients
```

The network protocol needs to identify which changes are persistent and which are ephemeral. On receiving a change over the network, a client applies persistent changes to both documents, and ephemeral changes only to the ephemeral document. Only the persistent document gets written to disk; the ephemeral document can be thrown away.

This approach should work well if the persistent and ephemeral changes update different properties. If the same property is independently updated in both the persistent and the ephemeral instance, this approach will result in a conflict. An alternative would be to discard the ephemeral document every time a persistent change is made, and re-initialise the ephemeral document as a clone of the persistent one:

```js
persistent = Automerge.change(persistent, doc => doc.persist = 'this!')
const change = Automerge.getLastLocalChange(persistent)
ephemeral = Automerge.clone(persistent)
// send change over the network to all other clients
```

That will avoid creating any conflicts, but it also discards any properties that have been added in the ephemeral document, so those would need to be re-created as an additional ephemeral change.

**Comment by skokenes** [comment]
This solution doesn't solve the history problem though, right? The ephemeral document is still keeping history around as long as its being used in memory. For something like cursors, the history can grow pretty rapidly. So is the idea that clients using ephemeral documents need to sometimes "reset" themselves by reinitializing their documents with `Automerge.init()`?

**Comment by ept** [comment]
At any time you can do `Automerge.free(ephemeral); ephemeral = Automerge.clone(persistent)`, which will throw away the state and history of the ephemeral document. (The freeing is only needed if using the webassembly backend, which requires memory to be freed explicitly when no longer needed.) However, after cloning, any state that should be present in the ephemeral document needs to be recreated. This is something to improve on when we add an explicit API for ephemeral changes.

**Comment by skokenes** [comment]
`free`/`clone` are just in 1.0 branch, right? Not in `0.14`?

**Comment by ept** [comment]
That's correct — sorry, I should have made that clear. In 0.x versions you can clone with `epemeral = Automerge.merge(Automerge.init(), persistent)`.

**Comment by dan-weaver** [comment]
How would nodes agree on when to free the ephemeral document?  That would have to be implemented within the networking layer?  I suppose there's multiple ways to do that but isn't it hard to know when all nodes are able to free their copy?

**Comment by ept** [comment]
The freeing can happen whenever the ephemeral document is cloned from the persistent one. When to clone, though? After every persistent change, maybe? But then there might be concurrent ephemeral changes from another user that would get thrown away. I don't really have a good answer for this at the moment. When we add explicit support for ephemeral updates, we will need to figure this out. The two-document approach is just an incomplete workaround for now.

**Comment by neftaly** [comment]
Any chance you'd consider adding a "good enough" solution to [@automerge/automerge-repo](https://github.com/automerge/automerge-repo)?

**Comment by pvh** [comment]
There's already an ephemeral data system in automerge-repo :) You can send an arbitrary message and it will be forwarded on to anyone else currently replicating that document.

**Comment by neftaly** [comment]
Oh thats awesome, sorry I have only been following this issue and hadn't seen that!

**Comment by neftaly** [comment]
Ok, from my understanding, to use Ephemeral Data:
* Create a regular repo, with a channelname prefixed with "m/" [Repo L63](https://github.com/automerge/automerge-repo/blob/060031ad2a48dde645139c23fee51ec64e6a026c/packages/automerge-repo/src/Repo.ts#L63)
* Broadcast data (object/primitives) using `repo.ephemeralData.broadcast(channelId, data)`
* Subscribe using `repo.ephemeral.on("data", ({ peerId, channelId, data }) => ())` (note that the "m/" channel prefix will be automatically removed)

Will ephemeral data propogate to other peers? Can it be sent to individual peers? [EphemeralData L16](https://github.com/automerge/automerge-repo/blob/main/packages/automerge-repo/src/EphemeralData.ts#L16)

edit: https://github.com/automerge/automerge-repo/pull/45
